test/unit/uk/gov/hmrc/exports/services/NotificationServiceSpec.scala

```diff
@@ -175,12 +182,62 @@ class NotificationServiceSpec extends WordSpec with MockitoSugar with ScalaFutur
         verify(submissionRepositoryMock, times(1)).updateMrn(meq(actionId), meq(mrn))
       }
 
-      trait SaveHappyPathTest extends Test {
-        when(notificationRepositoryMock.save(any())).thenReturn(Future.successful(true))
-        when(submissionRepositoryMock.updateMrn(any(), any())).thenReturn(Future.successful(Some(submission)))
+      "call NotificationRepository but not SubmissionRepository when saving an unparsed notification" in new SaveHappyPathTest {
+        notificationService.save(notificationUnparsable).futureValue
+
+        verify(notificationRepositoryMock, times(1)).save(meq(notificationUnparsable))
+        verifyNoInteractions(submissionRepositoryMock)
       }
     }
 
+    "NotificationService on reattemptParsingUnparsedNotifications" should {
+      "do nothing if no unparsed notifications exist" in new SaveHappyPathTest {
+        when(notificationRepositoryMock.findUnparsedNotifications())
+          .thenReturn(Future.successful(Seq(notification, notification_2)))
+
+        notificationService.reattemptParsingUnparsedNotifications().futureValue
+
+        verify(notificationRepositoryMock, times(0)).removeUnparsedNotificationsForActionId(any())
+        verify(notificationRepositoryMock, times(0)).save(any())
+      }
+
+      val dateTimeIssuedString = dateTimeIssued.format(ofPattern("yyyyMMddHHmmssX"))
+
+      "reparse single unparsed notification" in new SaveHappyPathTest {
+        val mrn = "GB1234567890"
+        val notificationNowParsable =
+          notificationUnparsable.copy(payload = NotificationTestData.exampleReceivedNotificationXML(mrn, dateTimeIssuedString).toString())
+
+        when(notificationRepositoryMock.findUnparsedNotifications())
+          .thenReturn(Future.successful(Seq(notificationNowParsable)))
+
+        notificationService.reattemptParsingUnparsedNotifications().futureValue
+
+        verify(notificationRepositoryMock, times(1)).removeUnparsedNotificationsForActionId(meq(notificationUnparsable.actionId))
+        verify(notificationRepositoryMock, times(1)).save(any())
+      }
+
+      "reparse single unparsed notification record (who's payload contains many 'responses')" in new SaveHappyPathTest {
+        val mrn = "GB1234567890"
+        val notificationNowParsable = notificationUnparsable.copy(
+          payload = NotificationTestData.exampleNotificationWithMultipleResponsesXML(mrn, dateTimeIssuedString).toString()
+        )
+
+        when(notificationRepositoryMock.findUnparsedNotifications())
+          .thenReturn(Future.successful(Seq(notificationNowParsable)))
+
+        notificationService.reattemptParsingUnparsedNotifications().futureValue
+
+        verify(notificationRepositoryMock, times(1)).removeUnparsedNotificationsForActionId(meq(notificationUnparsable.actionId))
+        verify(notificationRepositoryMock, times(2)).save(any())
```

Comment:
Nit: I like to use `org.mockito.InOrder` in case I want to ensure the order of calls (like here you want to save before removing).
Just saying ;-)

Reply:
Added